### PR TITLE
refactor: Remove `time.After` from any Loops

### LINKED ISSUE #14219: Remove time.After usages
from @christarazi :
```
❯ rg "<-time.After(.+)" --glob '!vendor' --glob '!*_test.go'
operator/kvstore_watchdog.go
88:                     <-time.After(defaults.LockLeaseTTL)
100:                    <-time.After(kvstore.HeartbeatWriteInterval)
operator/identity_gc.go
46:                     <-time.After(operatorOption.Config.IdentityGCInterval)
clustermesh-apiserver/main.go
543:                    <-time.After(kvstore.HeartbeatWriteInterval)
pkg/contexthelpers/context.go
37:             case <-time.After(timeout):
pkg/kvstore/etcd.go
759:            case <-time.After(initialConnectionTimeout):
1171:           case <-time.After(e.extraOptions.StatusCheckInterval(allConnected)):
pkg/hubble/relay/observer/observer.go
103:                    case t := <-time.After(bufferDrainTimeout):
pkg/hubble/relay/pool/manager.go
95:                     case <-time.After(m.opts.retryTimeout):
110:                    case <-time.After(m.opts.retryTimeout):
133:                            case <-time.After(m.opts.retryTimeout):
166:            case <-time.After(m.opts.connCheckInterval):
pkg/kvstore/consul.go
394:            case <-time.After(sleepTime):
pkg/k8s/watchers/watcher.go
327:            case <-time.After(option.Config.K8sSyncTimeout):
pkg/backoff/backoff.go
105:    case <-time.After(t):
pkg/kvstore/lock.go
107:            case <-time.After(time.Duration(10) * time.Millisecond):
pkg/kvstore/store/store.go
492:    case <-time.After(listTimeoutDefault):
pkg/trigger/trigger.go
216:            case <-time.After(t.params.sleepInterval):
pkg/node/manager/manager.go
308:            case <-time.After(syncInterval):
pkg/node/store/store.go
176:    case <-time.After(defaults.NodeInitTimeout / 10):
pkg/controller/controller.go
283:            case <-time.After(interval):
pkg/maps/ctmap/gc/gc.go
136:                    case <-time.After(ctmap.GetInterval(mapType, maxDeleteRatio)):
147:    case <-time.After(30 * time.Second):
proxylib/test/accesslog_server.go
65:             case <-time.After(10 * time.Millisecond):
pkg/status/status.go
156:                    case <-time.After(interval):
pkg/allocator/allocator.go
314:                    case <-time.After(listTimeout):
853:                    case <-time.After(option.Config.KVstorePeriodicSync):
pkg/rate/api_limiter.go
733:            case <-time.After(limitWaitDuration):
daemon/cmd/hubble.go
191:                            case <-time.After(30 * time.Second):
```

### PR DESCRIPTION
**Pull request by nathanjsweet**
`time.After` is not garbage collected until after
its channel fires once. In situations where it is called
repeatedely this *can* lead to significant memory build up.
Instead of creating a new timer object for every iteration,
a single timer, with a hard-reset mechanism, is used in
situations where `time.After` is being called repeatedely.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

Fixes: #14219

### CODE REVIEW

**Comment by brb** [comment]
Could we add a link to the article in the commit msg about time.After not being GC'd? It will help the future's readers.

**Comment by nathanjsweet** [comment]
test-me-please

File: pkg/inctimer/inctimer.go
Comment (aditighag) [inline pkg/inctimer/inctimer.go]: I'm not sure why we need a wrapper. Why isn't the in-place fix mentioned in the medium article sufficient?

Comment (christarazi) [reply pkg/inctimer/inctimer.go]: +1

**Review by christarazi — COMMENTED** [review]
The code looks good, but I'm also wondering the same as Aditi. Do we need a wrapper? If so, that would be helpful context in the commit msg and in the godoc of the `inctimer` package itself, for why it's necessary.

File: pkg/inctimer/inctimer.go
Comment (nathanjsweet) [reply pkg/inctimer/inctimer.go]: The solution suggested in the article is actually partially incorrect. In order to reset a timer  safely (as suggested in the article), you have to also [stop the timer and clear the channel](https://golang.org/pkg/time/#Timer.Reset).
I think this wrapper is useful because it basically boils the following code:
```golang
t := time.NewTimer(time.Nanosecond)
defer t.Stop()
for {
...
        if !t.Stop() {
		// It could be that the channel was read already
		select {
		case <-t.C:
		default:
		}
	}
        t.Reset(someInterval)
        select {
        ...
        case <- t.C:
        }
}
```
to:
```golang
t, stop := inctimer.New()
defer stop()
for {
...
        select {
        ...
        case <- t.After(someInterval):
        }
}
```
In one case I believe the `After` function is called even more than once, so that would have either required the creation of two separate timers or the same hard reset logic repeated twice above the select.
I find this wrapper approach to be a lot more elegant and easier to use.

**Review by christarazi — APPROVED** [review]
Awesome, looks great! Thanks for adding the extra context. I think it makes sense to have the wrapper. 

Thinking a bit into the future, I wonder how we can prevent uses of `time.After` without requiring code reviewers to remember. Help me understand, should we be replacing all uses of `time.After` with `IncTimer` going forward? If so, then we could do similar to what we have for `lock.Mutex` where we don't allow the use of the upstream `sync.Mutex` via a script (`contrib/scripts/lock-check.sh`) and this is enforced via our Golang GH Actions job.

**Comment by nathanjsweet** [comment]
> should we be replacing all uses of time.After with IncTimer going forward?
I think that's a little draconian. I think the rule should be any `time.After` use in a loop or recursive context should be replaced with `inctimer`.
> I wonder how we can prevent uses of time.After without requiring code reviewers to remember
This is a good question. I'll look into your `lock-check` example to see how it works.

**Review by aanm — CHANGES_REQUESTED** [review]
What's the rationality for having 2 different implementations, `pkg/status/status.go` vs `pkg/trigger/trigger.go`?

**Comment by nathanjsweet** [comment]
@aanm good catch. That's a pure mistake.

**Comment by nathanjsweet** [comment]
test-me-please

**Review by aditighag — APPROVED** [review]
Thanks for adding a custom vet tool too! :rocket:

File: clustermesh-apiserver/main.go
Comment (fristonio) [inline clustermesh-apiserver/main.go]: I suppose this should be a channel receive?

```suggestion
			<-timer.After(kvstore.HeartbeatWriteInterval)
```

**Review by fristonio — APPROVED** [review]
LGTM, thanks :rocket: Just one question/suggestion, does not need a second review.
Really cool work on the `customvet` check :tada:

**Comment by nathanjsweet** [comment]
test-me-please

**Review by kkourt — APPROVED** [review]
:rocket:

**Review by joestringer — APPROVED** [review]

**Review by jrajahalme — APPROVED** [review]

**Comment by tklauser** [comment]
This seems to have broken CI, see e.g. https://jenkins.cilium.io/job/cilium-ginkgo/job/cilium/job/master/6273

```
10:58:36      runtime: contrib/scripts/custom-vet-check.sh
10:58:36      runtime: # cd .; git clone -- https://github.com/cilium/customvet /home/vagrant/go/src/github.com/cilium/customvet
10:58:36      runtime: fatal: could not create work tree dir '/home/vagrant/go/src/github.com/cilium/customvet': Permission denied
10:58:36      runtime: package github.com/cilium/customvet: exit status 128
10:58:41      runtime: flag provided but not defined: -timeafter.ignore
10:58:41      runtime: usage: go vet [-n] [-x] [-vettool prog] [build flags] [vet flags] [packages]
10:58:41      runtime: Run 'go help vet' for details.
10:58:41      runtime: Run 'go tool vet -help' for the vet tool's flags.
10:58:41      runtime: Makefile:506: recipe for target 'precheck' failed
10:58:41      runtime: make: *** [precheck] Error 2
```

Looks like the CI runs failed on this PR already 😞

**Comment by tklauser** [comment]
I've sent a revert for this to unbreak CI: #14371

**Pull request merged**
